src/harness/test_harness_objects.py

```diff
@@ -0,0 +1,299 @@
+                                           grant_object)
+      return heartbeat_requests, heartbeat_responses
+
+  def mapResponseCodeToGrantState(self, heartbeat_response_code, grant_object):
+    """Maps the heartbeat response and update the grant state"""
+    if heartbeat_response_code == 502:
+      grant_object.is_terminated = True
+    elif heartbeat_response_code == 500:
+      grant_object.is_terminated = True
+      grant_object.authorized_in_last_heartbeat = False
+    elif heartbeat_response_code == 501:
+      grant_object.authorized_in_last_heartbeat = False
+
+  def performHeartbeatAndUpdateGrants(self):
+    """Does heartbeat request for all active grants and updates the grants.
+
+    The following steps are performed by this function
+      1. Send heartbeat request on all grant of CBSD at once
+      2. If the heartbeat response contains operationParam then those grants
+         are relinquished.
+      3. New grant requests with the suggested operationParam values in the
+         heartbeat response is sent. Failures are ignored.
+      4. Heartbeat is performed again for all grants. Failures are ignored.
+    """
+    # Heartbeat on all active grants
+    heartbeat_requests = []
+    heartbeat_requests, heartbeat_responses = self.heartbeatRequestForAllGrants()
+    relinquish_requests = []
+    grant_requests = []
+
+    # Construct relinquishment and grant for the heart beat response which have operation param
+    for heartbeat_response, heartbeat_request in zip(heartbeat_responses, heartbeat_requests):
+        if heartbeat_response['response']['responseCode'] != 0:
+          # If any heartbeats contain suggested operation params,construct relinquish requests
+          if 'operationParam' in heartbeat_response:
+            relinquish_request = {
+                'cbsdId': heartbeat_response['cbsdId'],
+                'grantId': heartbeat_response['grantId']
+            }
+            relinquish_requests.append(relinquish_request)
+
+            # Construct grant request for relinquished grants
+            grant_request = {
+              'cbsdId': heartbeat_request['cbsdId'],
+              'operationParam': {
+                'maxEirp': heartbeat_request['operationParam']['maxEirp'],
+                'operationFrequencyRange': {
+                  'lowFrequency': heartbeat_request['operationParam']['operationFrequencyRange']['lowFrequency'],
+                  'highFrequency': heartbeat_request['operationParam']['operationFrequencyRange']['highFrequency']
+                }
+              }
+            }
+            grant_requests.append(grant_request)
+
+          del self.cbsd_objects[heartbeat_response['cbsdId']]
+
+    # Perform relinquishment since operation param present in heartbeat response
+    if len(relinquish_requests):
+      relinquish_requests_wrap = {
+        'relinquishmentRequest': relinquish_requests
+      }
+      self.testcase._sas.Relinquishment(relinquish_requests_wrap,
```

Comment:
If any relinquishments fail it is most likely an issue with the test code (assuming the SAS UUT passes the basic protocol tests), regardless the tests and SAS UUT are in an inconsistent state so best to log what happened, and raise an error to stop the test execution (the test results would not be valid anymore anyway).

Reply:
Error   logged in case of  Relinquishment   fails